### ISSUE TITLE
Embedding a stand alone native app to my ionic 3 app

### ISSUE THREAD

**Issue by SibabratSwain** [mfdeveloper/cordova-plugin-nativeview#40]
I have an **ionic 3 app**. I needed to add video chat functionality to this existing app. I had achieved it using **webrtc** in mobile safari. I am using agora for the service. But in the recent iOS updated it does not support the streaming in safari. So I have built a separate native app for streaming. But I don't want to deploy two apps to AppStore in this situation rather embed the native caller app to the existing ionic app. 

I need to put the **channel id** in a form and submit in the main app which will open the embed native app and make the call. Then while I disconnect it will return to the existing ionic app. 

Could you please lighten me up in this situation as I have been searching so much and Found this plugin which seems to be working as per my purpose. 

**Ionic app folder structure**
![Ionic app folder structure](https://user-images.githubusercontent.com/38970037/89181940-45ca4400-d5b2-11ea-88e2-277a9459508c.png)

**Custome native app which I want to embed with the ionic app and can open with a button click from main ionic app**
![Screen Shot 1942-05-12 at 5 56 27 PM](https://user-images.githubusercontent.com/38970037/89182105-abb6cb80-d5b2-11ea-9edb-bb376b72856e.png)

I am really waiting for some suggestions and workaround for this situation.

**Comment by mfdeveloper** [comment]
Hy @SibabratSwain, how are you?

Sorry for my late response. I wait to the weekend to review issues here, because I've working during the week.

Do you still need some help for solve your scenario?

**Comment by SibabratSwain** [comment]
Hello @mfdeveloper. First of all, thank you so much for this plugin. As per my requirement, I wanted to open a native view controller from my Main Ionic app. I was so clueless about how to add the native app files inside the ionic main app. But finally, I have achieved that. Then using this plugin on a button click I am able to open the native screen pretty well. 

I have three measure issues now, please help me out. 

1 How to send a token as a parameter from web view to native view and catch it there. Because I need a run time token to be passed to execute some stuff on the native page. 
2 How to go back to the screen where it is opened from through e native button click event?
3 While it opened the native view I don't want to show the back button on the top left, How to remove that?

**Comment by SibabratSwain** [comment]
![Screen Shot 1942-05-16 at 4 06 32 PM](https://user-images.githubusercontent.com/38970037/89637562-fb093e80-d8c7-11ea-84fa-d33ee7e4d31a.png)

` cordova.plugins.NativeView.show('Main', 'ViewController').then().catch()`
I am using this to open the `ViewControler`, then from there I navigate to `VideoChatViewController`.

In the above screenshot, you can see a static variable channelID, I want to get that from the webView in params. 

and from a function inside the `VideoChatViewController` I want to return to the webview again.

**Comment by mfdeveloper** [comment]
Hello @SibabratSwain !!

How did you add the `VideoChatViewController.swift` to your IONIC project? If you puted him under `platforms/ios` folder, will be removed if you `ionic platforms ...` commands. I would suggest your add this like a [Cocoa pods](https://cocoapods.org/) dependency, or any other native dependency approach.

So, about your **"issues"**:

1 - Exists an opened issue #34 about pass parameters between `Cordova app` =>`Native view`. By now, just **Android** version was implemented. For IOS is not possible now, but I will do this in future. Do you have some "emergengy" to do this? I don't know whether you can wait for more days/weeks. Anyway, you can implement this and send a PR for me, and I can check, test and accept 😄 

2 - If you are using a `navigationController` (like a saw in the image above), in the `VideoChatViewController` you can do something like this:

**Go back to previous view**
```swift
self.navigationController?.popViewController(animated: true)

```
**Or go back to the "root" viewController**:  Defined in your Storyboard (in Cordova/Ionic app, is MainViewController, commonly)

```swift
self.navigationController?.popToRootViewController(animated: true)
```

3. You can create another `ViewController` that inherits from `MainViewController` and change this config on `viewDidLoad()` method:

```swift
self.navigationController.navigationBarHidden = true;

// Or this
self.navigationController.hidesBackButton = true;
```
Or, I can add a configuration with a <tag> in **`config.xml`**. In every cordova plugin, you can access `navigationController` and to anything. Maybe can be a good feature 😄 (But I don't have time to do this, now)

**Comment by SibabratSwain** [comment]
Hello @mfdeveloper. 
Though I am not an iOS developer I have somehow achieved to manage to pass the token from Cordova plugin to the ViewComtroller. I have achieved that using `UserDefaults`. 

**For my query number - 1** 
I have tried to initialize the swift class in the main file(CDVNativeView.m). But unable to do that also I am having less time to make it dynamic but I will work on it. 

@mfdeveloper I was thinking to make the plugin in swift, Which I will do and let you know. In that way, we can call the viewcontroller. 

**my query number - 3** 
Your solution worked fine. 

**my query number - 2** 
But your solution did not work. It shows a white screen. I need to close the app and remove it from memory to make it work again. 
Could you analyze more and give a solution to that. SO that it will go back to the previous screen where it redirected from.

**Comment by mfdeveloper** [comment]
Hello @SibabratSwain !!

So, about the `UserDefaults`, it's not "a standard way" in this case. Anyway, this implementation can work to you, and in the future I will implement send parameters, ok?

*my query number - 2*

Why did you generated ionic cordova project in `platforms` directory not contains a `navigationController` ? I was thinking that is a default in cordova projects. Maybe can be the **`cordova-ios`** version, I don't remember now 😞 

Can you send me an example project? Could be a example project with generic classes (whether you can't send me your business class) and attach here in this issue, or you can send me in private by e-mail, invite for another private github repository or somethingelse (which will be better for you)

Don't worry, I will help you and together, solve your scenario ok? 😃

**Comment by SibabratSwain** [comment]
@mfdeveloper Hello, how are you doing!

`self.dismiss(animated: true);`
Unfortunately did not work for me. 


I have the same issue mentioned in [https://github.com/mfdeveloper/cordova-plugin-nativeview/issues/8](#8). But he has fixed it with a patch. I am getting confused about where to add this to the plugin?

![Screen Shot 1942-05-26 at 3 49 06 PM](https://user-images.githubusercontent.com/38970037/90385908-68278b80-e0a1-11ea-81b5-265ae671cac8.png)

**Comment by mfdeveloper** [comment]
Hello @SibabratSwain !! I'm completely awesome, and you? 😃 

So, I don't know nothing about this patch. @krishnaclouds not send any PR to this repo 😢 I think that him implemented this in a fork. Ask for him directly in private, ok?

Sure!! I'm waiting for your invite to a github repo. By now, I don't have a Macos installed in my PC. I will do a Hackintosh again, but I need perform a whole backup of my HD, install the S.O and etc...

I will help you, but maybe not so fast. So sorry about this 😅

**Comment by SibabratSwain** [comment]
@mfdeveloper Hello! How are you doing? I have fixed the issue. As @krishnaclouds had suggested no need to set `rootViewController`. Rather I have changed the plugin a little bit which solved my issue. 

`self.viewController.view.window.rootViewController = nav;`
TO 
`[self.viewController.view.window.rootViewController presentViewController:nav animated:NO completion:nil];`

and ` self.dismiss(animated: true);` this works and I can redirect to ionic app from native view. Thanks, Man for the plugin.
